EIPS/eip-4337.md

```diff
@@ -0,0 +1,313 @@
+| `maxPriorityFeePerGas` | `uint256` | Maximum priority fee per gas (similar to EIP 1559 `max_priority_fee_per_gas`) |
+| `paymaster` | `address` | Address sponsoring the transaction (or zero for regular self-sponsored transactions) |
+| `paymasterData` | `bytes` | Extra data to send to the paymaster |
+| `signature` | `bytes` | Data passed into the wallet along with the nonce during the verification step |
+
+Users send `UserOperation` objects to a dedicated user operation mempool. A specialized class of actors called **bundlers** (either miners running special-purpose code, or users that can relay transactions to miners eg. through a bundle marketplace such as Flashbots) listen in on the user operation mempool, and create **bundle transactions**. A bundle transaction packages up multiple `UserOperation` objects into a single `handleOps` call to a pre-published global **entry point contract**.
+
+The core interface of the entry point contract is as follows:
+
+```c++
+function handleOps
+    (UserOperation[] calldata ops, address payable redeemer)
+    external;
+    
+function simulateWalletValidation
+    (UserOperation calldata userOp)
+    external returns (uint gasUsedByPayForSelfOp);
+    
+function simulatePaymasterValidation
+    (UserOperation calldata userOp, uint gasUsedByPayForSelfOp)
+    external view returns (bytes memory context, uint gasUsedByPayForOp);
+```
+
+The core interface required for a wallet to have is:
+
+```c++
+function validateUserOp
+    (UserOperation calldata userOp, uint requiredPrefund)
+    external;
+```
+
+### Required entry point contract functionality
+
+The entry point's `handleOps` function must perform the following steps (we first describe the simpler non-paymaster case). It must make two loops, the **verification loop** and the **execution loop**. In the verification loop, the `handleOps` call must perform the following steps for each `UserOperation`:
+
+* **Create the wallet if it does not yet exist**, using the initcode provided in the `UserOperation` (if the wallet does not exist _and_ the initcode is empty, the call must fail)
+* **Call `validateUserOp` on the wallet**, passing in the `UserOperation` and the required fee. The wallet should verify the operation's signature, and pay the fee if the wallet considers the operation valid. If any `validateUserOp` call fails, `handleOps` must skip execution of at least that operation, and may revert entirely.
+
+In the execution loop, the `handleOps` call must perform the following steps for each `UserOperation`:
+
+* **Call the wallet with the `UserOperation`'s calldata**. It's up to the wallet to choose how to parse the calldata; an expected worlflow is for the wallet to have an `execute` function that parses the remaining calldata as a series of one or more calls that the wallet should make.
+* **Refund unused gas fees** to the wallet
+
+![](https://storage.googleapis.com/ethereum-hackmd/upload_4b1c784e1563bb6ddf26886e86e59177.png)
```

Comment:
This file should be added to the PR. See EIP-1 for the exact directory structure, but I believe it should be stored in `/assets/eip-4337/image.png`.

Reply:
```suggestion
![](/assets/eip-4337/image1.png)
```